core/market/decentralized/tests/test_negotiations.rs

```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
 
     Ok(())
 }
+
+/// Can't counter proposal, that was already countered.
+/// Market should reject such attempts.
+#[cfg_attr(not(feature = "market-test-suite"), ignore)]
+#[actix_rt::test]
+async fn test_counter_countered_proposal() -> Result<(), anyhow::Error> {
+    let network = MarketsNetwork::new("test_counter_countered_proposal")
+        .await
+        .add_market_instance("Node-1")
+        .await?
+        .add_market_instance("Node-2")
+        .await?;
+
+    let market1 = network.get_market("Node-1");
+    let market2 = network.get_market("Node-2");
+
+    let identity1 = network.get_default_id("Node-1");
+    let identity2 = network.get_default_id("Node-2");
+
+    let demand_id = market1
+        .subscribe_demand(&sample_demand(), &identity1)
+        .await?;
+    let offer_id = market2.subscribe_offer(&sample_offer(), &identity2).await?;
+
+    // REQUESTOR side.
+    let proposal0 = requestor::query_proposal(&market1, &demand_id, 1).await?;
+    let proposal0_id = proposal0.get_proposal_id()?;
+
+    // Counter proposal for the first time.
+    let proposal1 = proposal0.counter_demand(sample_demand())?;
+    market1
+        .requestor_engine
+        .counter_proposal(&demand_id, &proposal0.get_proposal_id()?, &proposal1)
+        .await?;
```

Comment:
I'd add here some basic assertion to check if the first attempt was successful. Make sense?